src/lightning/pytorch/profilers/advanced.py

```diff
@@ -54,13 +58,17 @@ def __init__(
                 reported for each action. either an integer (to select a count of lines),
                 or a decimal fraction between 0.0 and 1.0 inclusive (to select a percentage of lines)
 
+            dump_stats: Whether to save raw profiler results. When ``True`` then ``dirpath`` must be provided.
+
         Raises:
             ValueError:
                 If you attempt to stop recording an action which was never started.
         """
         super().__init__(dirpath=dirpath, filename=filename)
         self.profiled_actions: Dict[str, cProfile.Profile] = {}
         self.line_count_restriction = line_count_restriction
+        self.dump_stats = dump_stats
+        assert not self.dump_stats or self.dirpath is not None, "dirname must be provided for dump_states to work"
```

Comment:
Could we make this a ValueError rather than an assertion error? That's because the error would be caused by user input.

Reply:
According to the docs:

```
If ``dirpath`` is ``None`` but ``filename`` is present, the
                ``trainer.log_dir`` (from :class:`~lightning.pytorch.loggers.tensorboard.TensorBoardLogger`)
                will be used.
```
So a None check here looks like is too early. We'd have to move it somewhere later after initialization.

Reply:
Ok, so `self.dirpath` is always available then.